1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+false && echo foo || echo bar # Writes only "bar" to stdout (operators are left-associative)
+true || echo foo && echo bar # Writes only "bar" to stdout
+```
+
+Also from the Shell Command Language specification:
+
+> A ';' or &lt;newline&gt; terminator shall cause the preceding AND-OR list to be executed sequentially;
+> an '&' shall cause asynchronous execution of the preceding AND-OR list.
+
+Meaning that the entire AND-OR list is backgrounded:
+
+```bash
+true && echo 'Success' & # The list `true && echo 'Success'` is executed in the background
+```
+
+This RFC proposes adding AND-OR lists to PowerShell, using the same `&&` and `||`.
+Because "AND-OR list" is neither intuitive nor common as a terminology,
+instead the term **pipeline chains** is proposed.
+
+## Motivation
+
+> As a PowerShell user, I can chain commands with `&&` and `||`
+> so that I have an ergonomic syntax to conditionally invoke side-effectful commands.
+
+## User Experience
+
+```powershell
+Write-Output "Hello" && Write-Output "Hello again"
+```
+
+```output
+Hello
+Hello again
+```
+
+```powershell
+Write-Output "Hello" || Write-Output "Hello again"
+```
+
+```output
+Hello
+```
+
+```powershell
+Write-Error "Bad" && Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+```
+
+```powershell
+Write-Error "Bad" || Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+
+Hello again
+```
+
+Using `echo` as a convenient example of a successful command
+and `false` as an example of an unsuccesful command.
+
+```powershell
+echo "Hello" && Write-Output "Hi"
+```
+
+```output
+Hello
+Hi
+```
+
+```powershell
+false && Write-Output "Reached"
+```
+
+```output
+```
+
+```powershell
+false || Write-Output "Reached"
+```
+
+```output
+Reached
+```
+
+```powershell
+false || Write-Output "Command failed" && Write-Output "Backup"
+```
+
+```output
+Command failed
+Backup
+```
+
+Also see: [intended test cases for implementation](https://github.com/PowerShell/PowerShell/blob/f8b899e42c86957a1b58273be0029f40a67bf1b6/test/powershell/Language/Operators/BashControlOperator.Tests.ps1).
+
+## Specification
+
+### Grammar
+
+Pipeline chains will be implemented using the already reserved `&&` and `||` operators.
+These operators will have the following grammar:
+
+```none
+statement:
+    | ... <existing statements except pipeline>   # Other statements
+    | variable_expression "=" statement           # Assignment
+    | pipeline_chain ["&"]                        # Pipeline chains
+
+pipeline_chain:
+    | pipeline
+    | pipeline_chain "&&" [newlines] pipeline
+    | pipeline_chain "||" [newlines] pipeline
+```
+

```

Comment:
This grammar should be amended to allow for automatic continuance when lines _start with_ `&&` and `||` as well, like so:

```none
pipeline_chain:
    | pipeline
    | pipeline_chain [newline] "&&" [newlines] pipeline
    | pipeline_chain [newline] "||" [newlines] pipeline
```

Note that when you are continuing a command with a pipeline on a subsequent line, only a single newline is allowed. This is consistent with how commands continue with a pipe symbol on a newline (a recent addition to PowerShell 7). Also note as an implementation detail that comment lines are allowed in the middle, but we don't specify those in the grammar. Additional test cases will need to be added for lines starting with these operators.

Further, while I did not change the multiple `newlines` after the operator to a single `newline` in this updated grammar, I believe we should consider that change. While PowerShell allows multiple newlines after a pipe or an operator, aside from supporting lines with comments, is there any real benefit in doing so?

Reply:
Wasn't aware that was in PowerShell -- will change

Reply:
Given that we decided to limit that type of lookahead multi-line continuance to just the pipeline operator, @PowerShell/powershell-committee agrees we should remove the `[newline]` from each of these. `[newlines]` after the operator are fine.